### PR TITLE
Fixed issue-4530: Added separate attestor type for secrets and KMS

### PR DESCRIPTION
**Pull request by shahpratikr**
Signed-off-by: Pratik Shah pratik@infracloud.io

## Explanation
The ImageVerify rule currently handles Secret integrations in the publicKeys field. Added separate attestor type for secrets which accepts name of the secret and namespace where it is created in fields.

## Related issue
Closes 4530

## What type of PR is this

/kind enhancement

## Proposed Changes

This PR adds ability to provide Kubernetes secret details as an attestor field.

```yaml
verifyImages:
- imageReferences:
  - "<image_name:tag>"
  attestors:
  - count: <count>
    entries:
    - keys:
        secret:
          name: "<secret_name>"
          namespace: "<namespace_name>"
    - keys:
        kms: gcpkms://...     
```

### Proof Manifests

Sign and push nginx image using following steps:
```bash
$ docker pull nginx:latest
$ docker tag nginx:latest pratikrshah/nginx:unsigned
$ docker push pratikrshah/nginx:unsigned
$ docker tag nginx:latest pratikrshah/nginx:cosign
$ docker push pratikrshah/nginx:cosign
For Secret:
$ cosign generate-key-pair k8s://default/test-secret
$ cosign sign --key k8s://default/test-secret docker.io/pratikrshah/nginx:cosign
For KMS:
$ cosign generate-key-pair --kms gcpkms://projects/<project_id>/locations/<location>/keyRings/<key_ring_name>/cryptoKeys/<key_name>/versions/<key_version>
$ cosign sign --key gcpkms://projects/<project_id>/locations/<location>/keyRings/<key_ring_name>/cryptoKeys/<key_name>/versions/<key_version> docker.io/pratikrshah/nginx:cosign-1
```
Policy:
```bash
apiVersion: kyverno.io/v1
kind: ClusterPolicy
metadata:
  name: check-image
  annotations:
    pod-policies.kyverno.io/autogen-controllers: none
spec:
  validationFailureAction: enforce
  background: false
  webhookTimeoutSeconds: 30
  failurePolicy: Fail
  rules:
  - name: check-image
    match:
      resources:
        kinds:
        - Pod
    verifyImages:
    - imageReferences:
      - "docker.io/pratikrshah/nginx:*"
      attestors:
      - count: 2
        entries:
        - keys:
            secret:
              name: "test-secret"
              namespace: "default"
        - keys:
            kms: gcpkms://projects/<project_id>/locations/<location>/keyRings/<key_ring_name>/cryptoKeys/<key_name>/versions/<key_version>
```
Create resource using unsigned tag:
```bash
$ kubectl run test --image docker.io/pratikrshah/nginx:unsigned
Error from server: admission webhook "mutate.kyverno.svc-fail" denied the request: 

policy Pod/default/test for resource violations: 

check-image:
  check-image: |
    failed to verify signature for docker.io/pratikrshah/nginx:unsigned: .attestors[0].entries[0].keys: no matching signatures:
verify-images: {}
```
Create resource using signed tag:
```bash
$ kubectl run test --image docker.io/pratikrshah/nginx:cosign
pod/test created
```

## Checklist
- [x] I have read the [contributing guidelines](https://github.com/kyverno/kyverno/blob/main/CONTRIBUTING.md).
- [x] I have read the [PR documentation guide](https://github.com/kyverno/kyverno/blob/main/.github/pr_documentation.md) and followed the process including adding proof manifests to this PR.
- [ ] This is a bug fix and I have added unit tests that prove my fix is effective.
- [ ] This is a feature and I have added CLI tests that are applicable.
- [ ] My PR needs to be cherry picked to a specific release branch which is <replace>.
- [ ] My PR contains new or altered behavior to Kyverno and
  - [ ] CLI support should be added and my PR doesn't contain that functionality.
  - [ ] I have added or changed [the documentation](https://github.com/kyverno/website) myself in an existing PR and the link is:
  
  
  - [ ] I have raised an issue in [kyverno/website](https://github.com/kyverno/website) to track the documentation update and the link is:

### CODE REVIEW

**Review by ApsTomar — CHANGES_REQUESTED** [review]

File: pkg/engine/imageVerify.go [hunk 1]
```diff
@@ -311,6 +311,8 @@ func (iv *imageVerifier) verifyAttestorSet(attestorSet kyvernov1.AttestorSet, im
 	var errorList []error
 	verifiedCount := 0
 	attestorSet = expandStaticKeys(attestorSet)
+	attestorSet = expandSecretData(attestorSet)
+	attestorSet = expandKmsData(attestorSet)
```
[the remark this answers is not in the document]
Comment (shahpratikr) [reply pkg/engine/imageVerify.go:315]: there might be the case where public keys are provided along with secret and kms. hence we need to keep all of those. these functions are handling public keys, secret and kms in different functions. it won't be a good idea to call `expandSecretData` and `expandKmsData` from `expandStaticKeys`. however, we can create wrapper function to call these and add a single call to that function here.

File: pkg/engine/imageVerify.go [hunk 2]
```diff
@@ -425,7 +457,14 @@ func (iv *imageVerifier) buildOptionsAndPath(attestor kyvernov1.Attestor, imageV
 
 	if attestor.Keys != nil {
 		path = path + ".keys"
-		opts.Key = attestor.Keys.PublicKeys
+		if attestor.Keys.PublicKeys != "" {
```
[the remark this answers is not in the document]
Comment (shahpratikr) [reply pkg/engine/imageVerify.go:460]: issue with using switch here is compare type is not same for all. also, we are comparing different fields for `attestor.Keys`

File: pkg/engine/imageVerify.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (ApsTomar) [reply pkg/engine/imageVerify.go:315]: In that case, yes a wrapper function would be nice. This function which will deal with all three expand calls.

**Review by ApsTomar — APPROVED** [review]
From code perspective, looks good.

**Review by samkulkarni20 — APPROVED** [review]

**Pull request merged**